src/Chainweb/Pact/PactService.hs

```diff
@@ -1269,66 +1301,75 @@ rewindTo
+            <> "; last hash: " <> blockHashToText (_blockHash lastHeader)
+            <> "; target height: " <> sshow parentHeight
+            <> "; target hash: " <> blockHashToText parentHash
+
+        failOnTooLowRequestedHeight rewindLimit lastHeader
+        playFork lastHeader
+
   where
-    rewindGenesis = return ()
-    doRewind (reqHeight, parentHash) = do
-
-        -- skip if the checkpointer is already at the target.
-        (_, lastHash) <- getCheckpointer >>= liftIO . _cpGetLatestBlock >>= \case
-            Nothing -> throwM NoBlockValidatedYet
-            Just p -> return p
-
-        unless (lastHash == parentHash) $ do
-
-            lastHeader <- findLatestValidBlock >>= maybe failNonGenesisOnEmptyDb return
-            logInfo $ T.unpack $ "rewind from last to target"
-                <> ". last height: " <> sshow (_blockHeight lastHeader)
-                <> "; last hash: " <> blockHashToText (_blockHash lastHeader)
-                <> "; target height: " <> sshow (reqHeight - 1)
-                <> "; target hash: " <> blockHashToText parentHash
-
-            failOnTooLowRequestedHeight rewindLimit lastHeader reqHeight
-            payloadDb <- asks _psPdb
-            bhDb <- asks _psBlockHeaderDb
-            playFork bhDb payloadDb parentHash lastHeader
-
-    failOnTooLowRequestedHeight (Just limit) lastHeader reqHeight
-      | reqHeight + limit < lastHeight = -- need to stick with addition because Word64
-        throwM $ RewindLimitExceeded
-        ("Requested rewind exceeds limit (" <> sshow limit <> ")")
-        reqHeight lastHeight
+    parentHeight = _blockHeight parent
+    parentHash = _blockHash parent
+
+    failOnTooLowRequestedHeight (Just limit) lastHeader
+        | parentHeight + 1 + limit < lastHeight = -- need to stick with addition because Word64
+            throwM $ RewindLimitExceeded
+                ("Requested rewind exceeds limit (" <> sshow limit <> ")")
+                parentHeight
+                lastHeight
       where
         lastHeight = _blockHeight lastHeader
-    failOnTooLowRequestedHeight _ _ _ = return ()
+    failOnTooLowRequestedHeight _ _ = return ()
 
     failNonGenesisOnEmptyDb = error "impossible: playing non-genesis block to empty DB"
 
-    playFork bhdb payloadDb parentHash lastHeader = do
-        parent <- ParentHeader <$!> lookupBlockHeader parentHash "rewindTo"
-
-        (!_, _, newBlocks) <-
-            liftIO $ collectForkBlocks bhdb lastHeader $ _parentHeader parent
-        -- play fork blocks
-        V.mapM_ (fastForward payloadDb) newBlocks
+    playFork lastHeader = do
+        bhdb <- asks _psBlockHeaderDb
+        (!commonAncestor, _, newBlocks) <-
+            liftIO $ collectForkBlocks bhdb lastHeader parent
+
+        if V.null newBlocks
+          then
+            -- If newBlocks is empty the checkpointer isn't restored via
+            -- 'fastForward'. So we do an empty 'withCheckPointerWithoutRewind'.
```

Comment:
Was this needed to fix something?

Reply:
Yes, this, along with 'initializeLatestBlock' is actually the main fix in this PR. For instance, 'initializeLatestBlock' alone doesn't fix the issue that restarts with an orphaned latest block fail, because `rewindTo` would be a noop, and the latest block would still point to the orphaned header. `_cpRestore' is only called in `withCheckpointerWithoutRewind`.